snuba/datasets/schemas/resolver.py

```diff
@@ -0,0 +1,126 @@
+from abc import ABC, abstractmethod
+from typing import Mapping, NamedTuple, Optional, Sequence
+
+from snuba.clickhouse.columns import ColumnSet
+from snuba.datasets.schemas.join import JoinNode
+
+
+class ResolvedCol(NamedTuple):
+    table_name: Optional[str]
+    column_name: str
+    path: Sequence[str]
+
+
+class ColumnResolver(ABC):
+    """
+    A ColumnResolver is a component that knows the logical schema of a data set and is capable
+    of resolving tables and nested columns from the representation we receive in the query.
+
+    This assumes the parser does not have enough information from the query language alone to
+    decompose a column expression found in a query into entity (table), base column name and
+    path for nested columns. Should we change the query language in such a way that a reference
+    to a column had non ambiguous information about entity name and nested columns this would
+    become a simple validator.
+
+    TODO: Revisit this interface when we introduce entities. We may have to increase its
+    responsibilities depending on how we decide to infer, given a column, which entity defines it.
+    If we require the query to fully qualify all columns (errors.message for example), this can
+    stay as it is. If instead we will infer the requested entity from the context of the query
+    like we do today this will need to know a lot more about the query, be coupled to that
+    structure and being potentially able to make wider changes to the query itself.
+    Moving such responsibility here would be easy enough that we can keep this simpler till
+    entities are introduced.
+    """
+
+    @abstractmethod
+    def resolve_column(self, query_column: str) -> Optional[ResolvedCol]:
```

Comment:
I wanted to decouple this class from the query language syntax, but that is fundamentally impossible since the behavior of this class depends on the syntax (for example the fact that a column can prepend the table name without saying it is a table name).